src/sentry/tagstore/v2/backend.py

```diff
@@ -412,9 +503,9 @@ def get_group_event_ids(self, project_id, group_id, environment_id, tags):
             EventTag.objects.filter(
                 project_id=project_id,
                 group_id=group_id,
-                key_id=k,
-                value_id=v,
-                **self._get_environment_filter(environment_id)
+                key_id__in=k,
+                value_id__in=v,
+                **env_filter
```

Comment:
I'd like to spend more time trying to figure out what kind of query plans this is going to generate in the `environment_id=None` case, unless you've already dug into this.

I don't think the `**env_filter` inclusion is necessary here since it's already been applied by limiting the `key` and `value` sets.  I don't expect that we would prefer to use the [`('environment_id', 'key', 'value')`](https://github.com/getsentry/sentry/blob/ab7e9381539ccc7e796f0fffa65014304b642715/src/sentry/tagstore/v2/models/eventtag.py#L33) index over the [`('group_id', 'key', 'value')`](https://github.com/getsentry/sentry/blob/ab7e9381539ccc7e796f0fffa65014304b642715/src/sentry/tagstore/v2/models/eventtag.py#L32) index, so this is just going to add at best an unnecessarily redundant filter condition, or at worst the wrong index, right? (I assume the `project_id` is necessary as a partition key, otherwise we could remove that as well.)

Another (more dubious) hunch: Since the `TagValue` objects are only associated with only one `TagKey` object, I'm suspicious that forcing the database to basically query for the Cartesian product of the `TagKey` and `TagValue` sets is going to be less efficient than writing a query that is more restricted to pairs that we already know are related.

Reply:
Damn, thunking face.

You're right about filtering the environment again, and about values being tied to a single key.

So why don't we just leave out both environment AND keys? Going to mess with this locally.